File: src/2025-03-04-12-20-21--deepseek-vr-0304/go/exercises/practice/dnd-character/dnd_character.go
package dndcharacter

import (
	"math/rand"
	"sort"
)

type Character struct {
	Strength     int
	Dexterity    int
	Constitution int
	Intelligence int
	Wisdom       int
	Charisma     int
	Hitpoints    int
}

// Modifier calculates the ability modifier for a given ability score
func Modifier(score int) int {
	return (score - 10) / 2
}

// Ability uses randomness to generate the score for an ability
func Ability() int {
	dice := make([]int, 4)
	for i := range dice {
		dice[i] = rand.Intn(6) + 1
	}
	sort.Ints(dice)
	return dice[1] + dice[2] + dice[3]
}

// GenerateCharacter creates a new Character with random scores for abilities
func GenerateCharacter() Character {
	c := Character{
		Strength:     Ability(),
		Dexterity:    Ability(),
		Constitution: Ability(),
		Intelligence: Ability(),
		Wisdom:       Ability(),
		Charisma:     Ability(),
	}
	c.Hitpoints = 10 + Modifier(c.Constitution)
	return c
}

File: src/2025-03-04-12-20-21--deepseek-vr-0304/cpp/exercises/practice/dnd-character/dnd_character.cpp
#include "dnd_character.h"
#include <vector>
#include <algorithm>
#include <random>
#include <cmath>

namespace dnd_character {

int ability() {
    static std::mt19937 gen(std::random_device{}());
    static std::uniform_int_distribution<> die(1, 6);

    std::vector<int> rolls;
    rolls.reserve(4);
    for (int i = 0; i < 4; ++i) {
        rolls.push_back(die(gen));
    }
    std::sort(rolls.begin(), rolls.end());
    return rolls[1] + rolls[2] + rolls[3];
}

Character::Character() {
    strength = ability();
    dexterity = ability();
    constitution = ability();
    intelligence = ability();
    wisdom = ability();
    charisma = ability();

    hitpoints = 10 + modifier(constitution);
}

int modifier(int constitution) {
    return static_cast<int>(std::floor((constitution - 10) / 2.0));
}

}  // namespace dnd_character
